text/0000-official-utility-buildpacks.md

```diff
@@ -0,0 +1,83 @@
+# Summary
+[summary]: #summary
+
+This is a proposal to adopt utility buildpacks as officially supported components of the Cloud Native Buildpacks project.
+
+# Definitions
+[definitions]: #definitions
+
+- **utility buildpack** - a buildpack that provides a generic capability that is useful across many vendors, platforms, and apps
+- **officially supported** - supported and maintained by the the Cloud Native Buildpacks project and its members
+
+# Motivation
+[motivation]: #motivation
+
+Many capabilities of lifecycle and pack do not require a native implementation within these components. These capabilities can (and have been) implemented at the user layer as buildpacks (ex. [inline-buildpack](https://registry.buildpacks.io/buildpacks/hone/inline)). In [PR #168](https://github.com/buildpacks/rfcs/pull/168) we assert that supporting these capabilities natively adds complexity to the implementation of lifecycle, which makes it difficult to innovate and improve.
+
+However, defering to the community and vendors to implement these kinds of utility buildpacks has not necessarily had good results. For example, there are three implementations of the [Procfile buildpack](https://registry.buildpacks.io/buildpacks/paketo-buildpacks/procfile). None of these implementations add significant (if any) advantages over the other. Yet buildpack users are force to decide which one to select. This is an unneccessary burden for a capability that could be provided by default.
+
+We seek to reduce complexity in the lifecycle and provide well maintained utility buildpacks for users by supporting them as official components of the Cloud Native Buildpacks project. This also aligns with our "everything's a buildpack" mantra.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Reserve the `cnb` namespace for officially supported buildpacks. Then, on a case by case basis, we adopt utility buildpacks under this
+
+Examples of potential utility buildpacks include:
+
+- `cnb/profile` - Evaluating `.profile` scripts in the app repo during launch
+- `cnb/inline` - Executing [Inline buildpacks](https://github.com/buildpacks/rfcs/blob/main/text/0048-inline-buildpack.md)
+- `cnb/processes` - Setting app specific process types from a configuration file like `project.toml`
+- `cnb/env` - Setting customer environment variables defined in a configuration file like `project.toml`
+- `cnb/purge` - Purging files from the app repo before export (similar to [`.slugignore`](https://devcenter.heroku.com/articles/slug-compiler#ignoring-files-with-slugignore))
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The utility buildpacks will be owned, supported, and maintained by the [Implementation team](https://github.com/buildpacks/community/blob/main/TEAMS.md#implementation-team).
+
+The criteria for accepting a utility buildpack into the project is as follows:
+- The behavior provided by the buildpack MUST be defined in the spec or an extension spec.
+- The buildpack MUST support arbitrary stacks
```

Comment:
Would this involve both linux and windows stacks? Would this also mean all the buildpacks would have to be compiled for all architectures?

Reply:
If we want the `.profile` buildpack to be the first utility buildpack, then I think this requirement is impossible. A `.profile` buildpack will never work on a stack that doesn't include Bash.

Reply:
I think we may need to let #172 finalize in order to address this. If `[[targets]]` and `os` under that becomes a thing, this will be much easier to deal with.

Reply:
What would that change L51 if that goes through?

Reply:
It depends on what we decide with #172. If wildcard targets are allowed, then these buildpacks probably use that. Otherwise, they'll have to specific all the targets they support explicitly.